### PR TITLE
feat(datepicker): year view, open on focus, bug fixes

### LINKED ISSUE #4251: md-datepicker should have option to select year and month
If we have to select future of past dates, (date of birth of something), it is difficult to scroll up and down. Should have selection of year and month.

### PR DESCRIPTION
**Pull request by crisbeto**
* Adds the ability for users to switch years quicker by clicking on the month header. Cheers to @mhchen for helping me get started.
* Adds the ability to open a datepicker by focusing the input.
* Fixes alignment issues, if the user has set the `box-sizing: border-box` on everything.
* Fixes cases where the calendar's width could be set to 0.

Fixes #4251, #4650, #8547, #8030, #8557.

### CODE REVIEW

**Comment by crisbeto** [comment]
@ThomasBurleson how should we proceed with reviewing this one? I had to shuffle around a lot of code in order to avoid duplication.

File: src/components/datepicker/calendar.js [hunk 1]
```diff
@@ -182,350 +104,173 @@
...
+      var value = this.$viewValue;
+
+      // Notify the child scopes of any changes.
+      self.$scope.$broadcast('md-calendar-parent-changed', value);
+
+      // Set up the selectedDate if it hasn't been already.
+      if (!self.selectedDate) self.selectedDate = value;
+
+      // Also set up the displayDate.
+      if (!self.displayDate) self.displayDate = self.selectedDate || self.today;
```
Comment (jelbourn) [inline src/components/datepicker/calendar.js]: Add braces

File: src/components/datepicker/calendar.spec.js [hunk 2]
```diff
@@ -66,16 +66,30 @@ describe('md-calendar', function() {
     return null;
   }
 
+  /** Find the `tbody` for a year in the calendar. */
+  function findYearElement(year) {
+    var years = element.querySelectorAll('[md-calendar-year-body]');
+    var yearHeader = year.toString();
```
Comment (jelbourn) [inline src/components/datepicker/calendar.spec.js]: Can we not stamp the year element with an attribute to more directly query it, similar to how the date cells operate?

Comment (jelbourn) [inline src/components/datepicker/calendar.spec.js]: Nevermind.

File: src/components/datepicker/calendarYear.js [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+  };
+
+  /**
+   * Hides the vertical scrollbar on the calendar scroller by setting the width on the
+   * calendar scroller and the `overflow: hidden` wrapper around the scroller, and then setting
+   * a padding-right on the scroller equal to the width of the browser's scrollbar.
+   *
+   * This will cause a reflow.
+   */
+  CalendarYearCtrl.prototype.hideVerticalScrollbar = function() {
```
Comment (jelbourn) [inline src/components/datepicker/calendarYear.js]: Can we refactor so that the year and month share code?

**Comment by jelbourn** [comment]
High level LGTM from me, thanks!

**Comment by crisbeto** [comment]
@ThomasBurleson @topherfangio @EladBezalel @DevVersion I've worked through the feedback from Jeremy. Can you give me some feedback on this?

File: src/components/datepicker/calendar.js [hunk 4]
```diff
@@ -182,350 +104,206 @@
...
     var self = this;
+
+    self.ngModelCtrl = ngModelCtrl;
+
     ngModelCtrl.$render = function() {
-      self.changeSelectedDate(self.ngModelCtrl.$viewValue);
+      var value = this.$viewValue;
+
+      // Notify the child scopes of any changes.
+      self.$scope.$broadcast('md-calendar-parent-changed', value);
```
Comment (EladBezalel) [inline src/components/datepicker/calendar.js]: I think we should avoid broadcasting, but if it's inevitable  I think it should be `md-calendar:parent-changed`

Comment (EladBezalel) [inline src/components/datepicker/calendar.js]: I think we should avoid broadcasting, but if it's inevitable  I think it should be `md-calendar:parent-changed`

You can avoid broadcasting by providing a `registerChanges` that you provide on the controller that gets a callback and saves it in a callback array and when the change happened go through the callback array and call it

File: src/components/datepicker/calendar.js [hunk 5]
```diff
@@ -182,350 +104,206 @@
...
    */
-  CalendarCtrl.prototype.buildInitialCalendarDisplay = function() {
-    this.buildWeekHeader();
-    this.hideVerticalScrollbar();
-
-    this.displayDate = this.selectedDate || this.today;
-    this.isInitialized = true;
+  CalendarCtrl.prototype.setNgModelValue = function(date) {
+    var value = this.dateUtil.createDateAtMidnight(date);
+    this.$scope.$emit('md-calendar-change', value);
```
Comment (EladBezalel) [inline src/components/datepicker/calendar.js]: Where we listen for this event?

File: src/components/datepicker/calendar.js [hunk 6]
```diff
@@ -182,350 +104,206 @@
...
+      case keyCode.END: return 'end';
 
-    var cellId = this.getDateId(date);
-    var cell = document.getElementById(cellId);
-    if (cell) {
-      cell.classList.add(FOCUSED_DATE_CLASS);
-      cell.focus();
-    } else {
-      this.focusDate = date;
+      default: return null;
```
Comment (EladBezalel) [inline src/components/datepicker/calendar.js]: You can just use return

File: src/components/datepicker/calendar.js (hunk 4 above)
Comment (crisbeto) [inline src/components/datepicker/calendar.js]: Doing a `registerChanges` method is basically the same as broadcasting it on the scope, except that we'd need to handle keeping track of the callbacks and cleaning up afterwards ourselves.

File: src/components/datepicker/calendar.js (hunk 5 above)
Comment (crisbeto) [inline src/components/datepicker/calendar.js]: It's used on parent scopes to know that the calendar changed. In this case it's only used in the datepicker, in order to know when to close the calendar.

**Comment by crisbeto** [comment]
It should be good to go now @ThomasBurleson.

**Comment by topherfangio** [comment]
Overall, I think this is really great and helpful!

A couple of thoughts from a UX perspective:

1. I think the year should have some additional feedback that it is clickable. Not sure the best way to do this, but maybe just changing the background on hover is enough (like we do with the days).

2. Without an animation between the two, it feels very jumpy and "un-material" IMO. Would it be fairly easy to add a slide or simple fade transition between the two?

3. Would it be possible to re-use the "day of the week" header area to provide a back button/icon? Right now, because the header disappears, it changes the size of the picker which feels odd, and we also don't give users an easy way to get back to the standard view if they decide they don't want to change anything after all.

Just some of my own personal thoughts 😄

**Comment by crisbeto** [comment]
@topherfangio 

1. Definitely, but I didn't find a nice way of doing it. [I've been trying stuff like this](http://i.imgur.com/mhTInNc.png), but I wouldn't say that it looks good. [I was leaning more towards something like this.](http://i.imgur.com/Z5IKIVA.png)

2. That sounds interesting! I couldn't get it to work nicely on the first try (for some reason the one with the datepicker didn't transition at all and the one outside a datepicker wasn't rendering the month content immediately), but I might be able to get it working.

3. I'm not sure about this one, that's a would be a pretty big back button.

**Comment by ThomasBurleson** [comment]
@crisbeto - Should i merge this now ?

**Comment by crisbeto** [comment]
I'd say so @ThomasBurleson. Once we have this set up, making improvements should be a bit easier.

**Pull request closed without merging**